crates/ntx-builder/src/builder/data_store/account_cache.rs

```diff
@@ -0,0 +1,288 @@
+                generation: 0,
+                accounts: HashMap::with_capacity(capacity),
+                ordering: BTreeMap::new(),
+            }),
+        }
+    }
+
+    /// Insert or replace an entry.
+    pub fn put(&self, account: &Account) {
+        let account_id = NtxAccountIdPrefix::from(account.id());
+        let mut state = self.state.lock().unwrap();
+        state.generation += 1;
+        let current_generation = state.generation;
+
+        let removed_generation = if let Some(entry) = state.accounts.insert(
+            account_id,
+            AccountEntry {
+                generation: current_generation,
+                account: account.clone(),
+            },
+        ) {
+            Some(entry.generation)
+        } else {
+            None
+        };
+
+        if let Some(old) = removed_generation {
+            state.ordering.remove(&old);
+        }
+        state.ordering.insert(current_generation, account_id);
+
+        if state.accounts.len() > self.capacity {
+            if let Some((_, oldest_key)) = state.ordering.pop_first() {
+                state.accounts.remove(&oldest_key);
+            }
+        }
+    }
+
+    /// Get a reference to a value if it exists in the cache.
+    /// Additionally, when there is a cache hit, marks the account as recently used.
+    pub fn get(&self, account_id: NtxAccountIdPrefix) -> Option<Account> {
+        let mut state = self.state.lock().unwrap();
+
+        let (old_generation, account) = {
+            state.generation += 1;
+            let new_generation = state.generation;
+            let entry = state.accounts.get_mut(&account_id)?;
+            let old = entry.generation;
+            entry.generation = new_generation;
+            (old, entry.account.clone())
+        };
+
+        let new_generation = state.generation;
+        state.ordering.remove(&old_generation);
+        state.ordering.insert(new_generation, account_id);
+
+        Some(account)
+    }
+
+    /// Manually evict an entry from the cache.
+    pub fn evict(&self, account_id: AccountId) -> Option<Account> {
+        let mut state = self.state.lock().unwrap();
+        if let Some(entry) = state.accounts.remove(&account_id.into()) {
+            state.ordering.remove(&entry.generation);
+            return Some(entry.account);
+        }
+        None
+    }
+}
+
+// TESTS
+// =================================================================================================
+
+#[cfg(test)]
+mod tests {
+    use std::sync::Arc;
+
+    use miden_lib::transaction::TransactionKernel;
+    use miden_objects::{Felt, account::Account};
+
+    use crate::builder::data_store::AccountCache;
+
+    #[test]
+    fn insert_get() {
+        let account = create_account(10);
+        let cache = AccountCache::new(2);
+
+        cache.put(&account.clone());
+        assert_eq!(cache.get(account.id().into()).unwrap(), account);
+    }
+
+    #[test]
+    fn lru_evicts_least_recently_used_account() {
```

Comment:
Any reason we wouldn't use something like https://docs.rs/lru/latest/lru/ instead of writing our own here?